SQL: Enable the InnerAggregates inside PIVOT

palesz opened the pull request:
* Remove the limitation of not being able use `InnerAggregate` inside
PIVOT (aggregations using extended and matrix stats)
* Integration tests that cover the extended stats and the matrix stats aggregations

Commit:
SQL: Enable the InnerAggregates inside PIVOT

* Remove the limitation of not being able use `InnerAggregate` inside
PIVOT (aggregations using extended and matrix stats)

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

astefan commented:
Why were the aggregates not supported inside PIVOT and what changed that made them now useable there?

palesz commented:
The limitation was introduced as part of the original `PIVOT` implementation: https://github.com/elastic/elasticsearch/pull/46489#issuecomment-532769953 , https://github.com/elastic/elasticsearch/pull/46489/commits/d6687510789e1d3e7a2838929eb6dd3997fedf74 .
The comment from @costin mentions that after the `NamedExpression` refactor (https://github.com/elastic/elasticsearch/pull/49693), probably should not be issues with the `InnerAggregates` inside `PIVOT`. 

Note: I will beef up the tests on this, and add extra, more complicated test-cases specifically mentioned in the original comment by @costin. For example: `ROUND(SUM_OF_SQUARES(...))`.

Commit:
Additional, more complicated testcases

costin reviewed (changes requested):
The reviewers list is incomplete

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/pivot.csv-spec:
```diff
@@ -201,6 +201,29 @@ null                 |10044          |Mingsen        |F              |1994-05-21
 
+sumWithInnerAggregateSumOfSquares
+schema::birth_date:ts|emp_no:i|first_name:s|gender:s|hire_date:ts|last_name:s|1:d|2:d
+SELECT * FROM test_emp PIVOT (SUM_OF_SQUARES(salary) FOR languages IN (1, 2)) LIMIT 5;
```
I'd like to see a test with multiple arguments that get folded into compound aggs, such as sum / min / max.

palesz replied:
Do you mean multiple aggregations inside the PIVOT? Currently there is a check that we can only have a single aggregation inside the `PIVOT`, so we can't have `PIVOT(MIN(salary), MAX(salary) FOR ...)`.

This change only lifts the limitation on the *type* of that single aggregation within the PIVOT, so aggregations that are translated into an `InnerAggregate` will be allowed.

costin replied:
Is there any validation or check for this (now that PostOptimizerTest has been removed)? this goes beyond the scope of this ticket so please raise a ticket if this functionality is missing or needs to be further investigated.

palesz replied:
The `PostOptimizerTests` checked that the *type* of the aggregation cannot be one that results in an InnerAggregate vs the number of the aggregations inside the PIVOT.

There is already a check in the `LogicalPlanBuilder` that only allows one aggregation inside the PIVOT: https://github.com/elastic/elasticsearch/blob/2f5ab81b9ad884434d1faa60f4391f966ede73e8/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java#L186

I cannot see it tested in unit tests, but it is there and works. Added: #65894

costin commented:
> Note: I will beef up the tests on this, and add extra, more complicated test-cases specifically mentioned in the original comment by @costin. For example: ROUND(SUM_OF_SQUARES(...)).

I'm in favor of removing code however when dealing with known issues, the tests are critical to assess that there are no regressions. In fact, for this type of PR they become the main part - without them this PR is incomplete.

Commit:
Test that `PIVOT` and `GROUP BY` has same query

Test that the `PIVOT` results in the same query as
the `GROUP BY`. This should hold across all the `AggregateFunction`s we
have.

palesz commented on x-pack/plugin/sql/qa/server/src/main/resources/pivot.csv-spec:
```diff
@@ -201,6 +201,51 @@ null                 |10044          |Mingsen        |F              |1994-05-21
+
+  client_port  |       'OK'       |    'Error'    
+---------------+------------------+---------------
+null           |2.002             |-0.0            
```
Created https://github.com/elastic/elasticsearch/issues/65863

Commit:
Merge remote-tracking branch 'origin/master' into feature/pivot-inneraggregates

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java:
```diff
@@ -495,14 +497,34 @@ public void testFoldingOfPivot() {
         assertThat(a, containsString("\"terms\":{\"field\":\"keyword\""));
         assertThat(a, containsString("{\"avg\":{\"field\":\"int\"}"));
     }
+    
```
It's better to get the list from the Function registry so that all existing aggs and future ones are picked up without having to manually update this test (since it is unlikely it will happen).

palesz replied:
:+1: Yep, that sounds like a better solution (was thinking about possible going with reflection, but this is better).

costin reviewed (commented):
Left another round of comments.

Commit:
Utilized SQLFunctionRegistry to look up the `AggregateFunction`s

Commit:
Merge remote-tracking branch 'origin/master' into feature/pivot-inneraggregates

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java:
```diff
@@ -495,14 +499,39 @@ public void testFoldingOfPivot() {
     }
+    
+    public void testPivotHasSameQueryAsGroupBy() {
+        final Map<String, String> aggFnsWithMultipleArguments = Map.of(
```
You might want to use a different construct to ease backporting to 7.x due to the JDK difference

palesz replied:
I was thinking about this. I would prefer using the new APIs on master and change during backporting to Java 8 compatible, so moving forward we use the newer APIs.

costin reviewed (approved):
LGTM

bpintea reviewed (approved):
LGTM

astefan commented on x-pack/plugin/sql/qa/server/src/main/resources/pivot.csv-spec:
```diff
@@ -201,6 +201,51 @@ null                 |10044          |Mingsen        |F              |1994-05-21
 
+sumWithInnerAggregateSumOfSquares
+schema::birth_date:ts|emp_no:i|first_name:s|gender:s|hire_date:ts|last_name:s|1:d|2:d
+SELECT * FROM test_emp PIVOT (SUM_OF_SQUARES(salary) FOR languages IN (1, 2)) LIMIT 5;
```
This query fails for me in a REST test (using a REST client): it's only returning 3 rows, whereas it should have returned 5. We need to look into why the test passes and/or why the REST call with the same query fails. @palesz please, open an issue for this investigation.

palesz replied:
The JDBC integration tests fetch all the rows and follow the cursor (hence 5 rows shows up there). The tests are valid and successful, no issue there. Created https://github.com/elastic/elasticsearch/issues/65982 to investigate this unexpected behaviour related to `PIVOT+LIMIT`.

astefan reviewed (commented):
There is one issue that needs further analysis (added a comment for pivot.csv-spec). Still reviewing.

astefan commented:
LGTM otherwise.
